### PR TITLE
LwM2M 1.10 changes: firmware support + bugfixes + more (2nd version)

### PR DESCRIPTION
**Pull request by mike-scott**
NOTE: Original PR (https://github.com/zephyrproject-rtos/zephyr/pull/4208) was corrupted and stopped responding to updates to the base branch in my repo.  This is a new version of that PR.

This patchset does the following:

support for firmware handling: including push and pull methods
moves some SYS_LOG message from DBG to INF for better visibility
couple of bugfixes which leaked packets
retry handling for firmware downloads
application callbacks from the registration client

### CODE REVIEW

**Comment by mike-scott** [comment]
Rebased against current master.  This PR is essentially https://github.com/zephyrproject-rtos/zephyr/pull/4208 except that the first commit was reworked.

The relevant diff from https://github.com/zephyrproject-rtos/zephyr/pull/4208 is here: https://hastebin.com/aluqigaker.hs

**Comment by SebastianBoe** [comment]
@mike-scott : Seems github is having some issues at the moment.

https://status.github.com/messages

**Comment by mike-scott** [comment]
@SebastianBoe Thank you!  I was still looking to see if something on my end was broken :)

File: include/net/lwm2m.h [hunk 1]
```diff
@@ -123,10 +123,10 @@ int lwm2m_device_add_err(u8_t error_code);
 #define RESULT_UNSUP_PROTO	9
```
Comment (jukkar) [inline include/net/lwm2m.h]: Is there a typo in commit message s/btw/between/ ?

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c [hunk 2]
```diff
@@ -162,6 +162,7 @@ do_firmware_transfer_reply_cb(const struct zoap_packet *response,
 	struct zoap_packet *check_response = (struct zoap_packet *)response;
```
Comment (jukkar) [inline subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c]: Typo in commit msg s/Both/both/

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c [hunk 3]
```diff
@@ -173,13 +174,26 @@ do_firmware_transfer_reply_cb(const struct zoap_packet *response,
...
 	if (ret < 0) {
 		SYS_LOG_ERR("Error from block update: %d", ret);
 		lwm2m_firmware_set_update_result(RESULT_INTEGRITY_FAILED);
 		return ret;
 	}
 
+	/* test for duplicate transfer */
+	if (firmware_block_ctx.current < received_block_ctx.current) {
+		SYS_LOG_WRN("Duplicate packet ignored");
+		/* restore main firmware block context */
```
Comment (jukkar) [inline subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c]: Perhaps empty line before comment

**Review by jukkar — CHANGES_REQUESTED** [review]

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c [hunk 4]
```diff
@@ -77,14 +76,43 @@ static int transfer_request(struct zoap_block_context *ctx,
...
+	/* if path is not available, off/len will be zero */
+	off = parsed_uri.field_data[UF_PATH].off;
+	len = parsed_uri.field_data[UF_PATH].len;
+	cursor = firmware_uri + off;
+	path_len = 0;
+
+	for (i = 0; i < len; i++) {
+		if (firmware_uri[off + i] == '/') {
+			if (path_len > 0) {
+				ret = zoap_add_option(&msg->zpkt,
```
[the remark this answers is not in the document]
Comment (mike-scott) [reply subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c]: New CoAP API changes are coming after this patchset.

**Comment by mike-scott** [comment]
Will address @jukkar comments.  Thanks for the review!

File: subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (mike-scott) [reply subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c]: @jukkar for reference here's my current patch queue for LwM2M / CoAP / DTLS (based on master): https://github.com/mike-scott/zephyr/commits/master-lwm2m

This patchset ends near the bottom of the first page with "net: lwm2m: add RD client callbacks for app".

**Comment by mike-scott** [comment]
Updated commit descriptions per @jukkar and here's the diff for the spacing change: https://hastebin.com/tajiqadumo.swift

**Review by jukkar — APPROVED** [review]

**Comment by mike-scott** [comment]
Quick Update: Found a left-over CONFIG name in the LwM2M Kconfig
Diff: https://hastebin.com/onegegacov.vbs

**Comment by nashif** [comment]
just fix that commit message line length...

**Review by nashif — APPROVED** [review]

**Comment by mike-scott** [comment]
@nashif Woops!  Re-pushed with the commit description line length fixed.  Also, double checked checkpatch on the other patches as well.  SHIPPABLE should pass now.

**Comment by mike-scott** [comment]
Ok all looks good now. :)

File: include/net/lwm2m.h (hunk 1 above)
Comment (edwardofclt) [reply include/net/lwm2m.h:123]: Confirmed pre-existing and out of scope for TEF-5688, which is scoped narrowly to replacing the hand-constructed `traceparent` with `injectTraceContext` (audit finding I-32). Subject whitelisting belongs in a separate follow-up audit ticket tracking `scaffolder-action-contracts.md` rule 3 compliance.

**Pull request merged**